main.nf

```diff
@@ -167,46 +486,759 @@ process fastqc {
 
     script:
     """
-    fastqc -q $reads
+    fastqc -t "${task.cpus}" -q $reads
     """
 }
 
 
+process fastp {
+    tag "$name"
+    publishDir "${params.outdir}/", mode: 'copy',
+        saveAs: {filename -> filename.indexOf(".fastq.gz") == -1 ? "QC_shortreads/fastp/$name/$filename" : null}
+
+    input:
+    set val(name), file(reads) from read_files_fastp
+    val adapter from params.adapter_forward
+    val adapter_reverse from params.adapter_reverse
+    val qual from params.mean_quality
+    val trim_qual from params.trimming_quality
+
+    output:
+    set val(name), file("${name}_trimmed*.fastq.gz") into trimmed_reads
+    file("fastp.*")
+
+    script:
+    if ( !params.singleEnd ) {
+        """
+        fastp -w "${task.cpus}" -q "${qual}" --cut_by_quality5 \
+            --cut_by_quality3 --cut_mean_quality "${trim_qual}"\
+            --adapter_sequence=${adapter} --adapter_sequence_r2=${adapter_reverse} \
+            -i "${reads[0]}" -I "${reads[1]}" \
+            -o ${name}_trimmed_R1.fastq.gz -O ${name}_trimmed_R2.fastq.gz
+        """
+    }
+    else {
+        """
+        fastp -w "${task.cpus}" -q "${qual}" --cut_by_quality5 \
+            --cut_by_quality3 --cut_mean_quality "${trim_qual}"\
+            --adapter_sequence="${adapter}" --adapter_sequence_r2="${adapter_reverse}" \
+            -i ${reads} -o "${name}_trimmed.fastq.gz"
+        """
+    }
```

Comment:
As this is mostly the same, it might be nicer to have a little groovy code above this which sets the additional command line optoins if needed. eg:

```nextflow
script:
def pe_input = params.singleEnd ? '' :  "-I \"${reads[1]}\""
def pe_output1 = params.singleEnd ? "-o \"${name}_trimmed.fastq.gz\"" :  "-o \"${name}_trimmed_R1.fastq.gz\""
def pe_output2 = params.singleEnd ? '' :  "-O \"${name}_trimmed_R2.fastq.gz\""
"""
fastp -w ... $pe_input $pe_output1 $pe_output2
"""
```

..etc

Reply:
This means that there is no duplication of the script block, so less vulnerable to copy pasta errors in the future

Reply:
done!